sky/spot/recovery_strategy.py

```diff
@@ -444,3 +450,77 @@ def recover(self) -> float:
                         f'{self._MAX_RETRY_CNT} times.')
 
             return job_submitted_at
+
+
+class EagerFailoverStrategyExecutor(FailoverStrategyExecutor,
+                                    name='EAGER_FAILOVER',
+                                    default=True):
+    """Aggressive failover strategy
+
+    This strategy is an extension of the failover strategy. Instead of waiting
+    in the same region when the preemption happens, it immediately terminates
+    the cluster and relaunches it in a different regions. This is based on the
+    observation that the preemption is likely to happen again shortly in the
+    same region, so trying other regions first is more likely to get a longer
+    running cluster.
+    """
+
+    def recover(self) -> float:
+        # 1. Terminate the current cluster
+        # 2. Launch the cluster without retrying the previously launched region
+        # 3. Launch the cluster with no cloud/region constraint or respect the
+        #    original user specification.
+
+        # Step 1
+        logger.debug('Terminating unhealthy spot cluster and '
+                     'reset cloud region.')
+        terminate_cluster(self.cluster_name)
+
+        # Step 2
+        logger.debug('Relaunch the cluster skipping the previously launched '
+                     'cloud/region.')
+        if self._launched_cloud_region is not None:
+            task = self.dag.tasks[0]
+            resources = list(task.resources)[0]
+            if resources.region is None and resources.zone is None:
+                # Optimization: We only block the previously launched region,
+                # if the task does not specify a region or zone, because,
+                # otherwise, we will spend unnecessary time for skipping the
+                # only specified region/zone.
+                launched_cloud, launched_region = self._launched_cloud_region
+                task.blocked_resources = {
+                    resources.copy(cloud=launched_cloud, region=launched_region)
+                }
+                # Not using self.launch to avoid the retry until up logic.
+                job_submitted_at = self._launch(raise_on_failure=False)
+                task.blocked_resources = None
```

Comment:
What's the difference b/t resetting it here vs. resetting when the underlying sky.launch() throws ResourcesUnavailableError (exhausted all launchables)?

I imagine that if we reset here, the following is possible. Assume the user has access to 3 regions, R1, R2, R3, in that price order:
```
R1Z1 (preempted) -> R2 (failed to launch; reset blocked_resources) -> R1Z1 (very likely to be preempted again)
```
If we reset in the second way, then we may have
```
R1Z1 (preempted) -> R2 (failed to launch; add R2 to blocked_resources) -> R3 (giving it a shot)
```

Reply:
Got it. Could we add this example to the class' docstr to remind ourselves?

```
Example: Assume the user has access to 3 regions, R1, R2, R3, in that price order. Then the following are some possible event sequences:

R1Z1 (preempted) -> R2 (success)
R1Z1 (preempted) -> R2 (failed to launch) -> R3 (success)
R1Z1 (preempted) -> R2 (failed to launch) -> R3 (failed to launch) -> R1Z2 (success)
R1Z1 (preempted) -> R2 (failed to launch) -> R3 (failed to launch) -> R1Z1 (success)
```
Or something better?